src/Lykke.Service.FixGateway.Services/MarketDataRequestHandler.cs

```diff
@@ -0,0 +1,280 @@
+
+        }
+
+        private async Task<bool> ValidateRequestAsync(MarketDataRequest request)
+        {
+            Message reject = null;
+            if (_subscriptions.ContainsKey(request.MDReqID.Obj))
+            {
+                reject = GetFailedResponse(request, MDReqRejReason.DUPLICATE_MDREQID);
+            }
+
+            if (request.SubscriptionRequestType.Obj != SubscriptionRequestType.SNAPSHOT_PLUS_UPDATES
+                && request.SubscriptionRequestType.Obj != SubscriptionRequestType.DISABLE_PREVIOUS_SNAPSHOT_PLUS_UPDATE_REQUEST)
+            {
+                reject = GetFailedResponse(request, MDReqRejReason.UNSUPPORTED_SUBSCRIPTIONREQUESTTYPE);
+            }
+
+            if (request.MarketDepth.Obj < 0)
+            {
+                reject = GetFailedResponse(request, MDReqRejReason.UNSUPPORTED_MARKETDEPTH);
+            }
+
+            var entryInvalid = false;
+            for (var i = 1; i <= request.NoMDEntryTypes.Obj; i++)
+            {
+                var entityType = ((MarketDataRequest.NoMDEntryTypesGroup)request.GetGroup(i, new MarketDataRequest.NoMDEntryTypesGroup())).MDEntryType;
+
+                if (entityType.Obj != MDEntryType.OFFER && entityType.Obj != MDEntryType.BID)
+                {
+                    entryInvalid = true;
+                    break;
+                }
+            }
+
+            if (entryInvalid)
+            {
+                reject = GetFailedResponse(request, MDReqRejReason.UNSUPPORTED_MDENTRYTYPE);
+            }
+
+            var allSymbols = (await _assetsServiceWithCache.GetAllAssetPairsAsync(_tokenSource.Token)).Where(ap => !ap.IsDisabled)
+                .Select(ap => ap.Id)
+                .ToHashSet();
+
+            var hasInvalidSymbol = false;
+            for (var i = 1; i <= request.NoRelatedSym.Obj; i++)
+            {
+                var symbol = ((MarketDataRequest.NoRelatedSymGroup)request.GetGroup(i, new MarketDataRequest.NoRelatedSymGroup())).Symbol.Obj;
+
+                if (!allSymbols.Contains(symbol))
+                {
+                    hasInvalidSymbol = true;
+                    break;
+                }
+            }
+
+            if (hasInvalidSymbol)
+            {
+                reject = GetFailedResponse(request, MDReqRejReason.UNKNOWN_SYMBOL);
+            }
+
+            if (reject != null)
+            {
+                Send(reject);
+                return false;
+            }
+            return true;
+        }
+
+        private static Message GetFailedResponse(MarketDataRequest request, char rejectReason)
+        {
+            var reject = new MarketDataRequestReject
+            {
+                MDReqID = request.MDReqID,
+                MDReqRejReason = new MDReqRejReason(rejectReason)
+            };
+            return reject;
+        }
+
+
+        private void Send(Message reject)
```

Comment:
maybe rename parameter reject to message, rly this method for send any message or need change type